client.go

```diff
@@ -0,0 +1,449 @@
+func WithRequestHeaders(headers http.Header) ClientStreamOption {
+	return &withRequestHeaders{headers: headers}
+}
+
+type withRequestHeaders struct {
+	headers http.Header
+}
+
+func (w *withRequestHeaders) apply(options *clientStreamOptions) {
+	options.headers = w.headers
+}
+
+// WithReflectionHost is an option that allows the caller to provide the hostname that
+// will be included with all requests on the stream. This may be used by the server
+// when deciding what source of reflection information to use (which could include
+// forwarding the request message to a different host).
+func WithReflectionHost(host string) ClientStreamOption {
+	return &withReflectionHost{host: host}
+}
+
+type withReflectionHost struct {
+	host string
+}
+
+func (w *withReflectionHost) apply(options *clientStreamOptions) {
+	options.host = w.host
+}
+
+// NewStream creates a new stream that is used to download reflection information from
+// the server. This is a bidirectional stream, so can only be successfully used over
+// HTTP/2. The [ClientStream.Close] method should be called when the caller is done
+// with the stream.
+//
+// If any operation returns an error for which [IsReflectionStreamError] returns true,
+// the stream is broken and all subsequent operations will fail. If the error is not
+// a permanent error, the caller should create another stream and try again.
+func (c *Client) NewStream(ctx context.Context, options ...ClientStreamOption) *ClientStream {
+	clientStream := &ClientStream{
+		ctx:    ctx,
+		client: c,
+	}
+	for _, option := range options {
+		option.apply(&clientStream.clientStreamOptions)
+	}
+	// warm-up the stream
+	clientStream.getStream()
+	return clientStream
+}
+
+// IsReflectionStreamError returns true if the given error was the result of a [ClientStream]
+// failing. If the stream returns an error for which this function returns false, only the
+// one operation failed; the stream is still intact and may be used for subsequent operations.
+func IsReflectionStreamError(err error) bool {
+	var streamErr *streamError
+	return errors.As(err, &streamErr)
+}
+
+// ClientStream represents a bidirectional stream for downloading reflection information.
+// The reflection protocol resembles a sequence of unary RPCs: multiple requests sent on the
+// stream, each getting back a corresponding response. However, all such requests and responses
+// and sent on a single stream to a single server, to ensure consistency in the information
+// downloaded (since different servers could potentially have different versions of reflection
+// information).
+type ClientStream struct {
+	ctx    context.Context //nolint:containedctx
+	client *Client
+	clientStreamOptions
+
+	mu     sync.Mutex
+	stream *reflectStream
+	isV1   bool
+}
+
+func (cs *ClientStream) getStream() *reflectStream {
+	cs.mu.Lock()
+	defer cs.mu.Unlock()
+	return cs.getStreamLocked()
+}
+
+func (cs *ClientStream) getStreamLocked() *reflectStream {
```

Comment:
:+1: Like that the function name clearly communicates that the caller must manage the mutex!